### ISSUE TITLE
Deadlock on main thread creating ANR in the app in MonitoringStatus.regions()

### ISSUE THREAD

**Issue by kevingilles** [AltBeacon/android-beacon-library#1108]
### Expected behavior
altbeacon would not block the main thread

### Actual behavior
I can find a deadlock on the mainThread that creates ANR. You can find the lock in the screenshot below.

### Steps to reproduce this behavior
I'm not sure, it's sporadic and I haven't reproduced it myself. II can see it in Firebase though.

### Mobile device model and OS version
Android 12 - Galaxy A32

### Android Beacon Library version
2.19.4

<img width="983" alt="Deadlock on my app" src="https://user-images.githubusercontent.com/45975964/197775377-8ff974d2-1dae-4b8d-8925-2a39af318872.png">

**Comment by davidgyoung** [comment]
Thanks for this report.  

I have been starting at the code for 30 minutes, but I can't see how a deadlock happened here.  AFAICT, here is what is happening when this happens:

The main thread received a new scan via intent, and tried to restore scan state to schedule a new scan job with this line `ScanState scanState = ScanState.restore(context);`, this caused it to get exclusive access to the `ScanState` object in this line:  `synchronized (ScanState.class)`.  It continues and inside that block, it ultimately wants to call this method `scanState.getMonitoringStatus().regions()` but gets blocked as it is synchronized`MonitoringStatus` class which is synchronized here: ` public synchronized Set<Region> regions() { return getRegionsStateMap().keySet(); }`

Meanwhile, another thread was trying to start beacon ranging, which also caused that thread to try to restore scan state with the same call to `ScanState scanState = ScanState.restore(context);`, it gets blocked on the `synchronized (ScanState.class)`


So in summary:

* The main thread is inside a synchronized block on `ScanState.class` and is blocked waiting to call a synchronized method on `MonitoringStatus`.
* A background thread is waiting to get synchronized access to the `ScanState.class`.  

Why is this a deadlock?  I see no reason why the main thread cannot call the method on `MonitoringStatus`.  The second thread is not accessing it.

Unless there is another thread in the mix not shown in the screenshot?

**Comment by kevingilles** [comment]
It is indeed eluding me too, thank you for your time. I managed to get my hands on 2 more exemples that I might help with context.
I hope this information is useful, I remain available to test anything on my end you would find relevant.

### Here is the first one: 
<img width="976" alt="Capture d’écran 2022-10-28 à 15 03 32" src="https://user-images.githubusercontent.com/45975964/198600224-a882003c-39e7-4297-8cf9-002e411b291a.png">

### And here is the second one (2 different Screenshots)
<img width="932" alt="Capture d’écran 2022-10-28 à 15 04 19" src="https://user-images.githubusercontent.com/45975964/198600305-54879437-da6e-42ef-a066-355d57714dce.png">
<img width="932" alt="Capture d’écran 2022-10-28 à 15 04 32" src="https://user-images.githubusercontent.com/45975964/198600353-83ecf672-3660-4b58-8a21-7f0f44f99790.png">

### Finally, the 3rd one
<img width="842" alt="Capture d’écran 2022-10-28 à 15 11 51" src="https://user-images.githubusercontent.com/45975964/198601570-17d58d63-45b1-42a4-a44d-8a18f90229d1.png">
<img width="842" alt="Capture d’écran 2022-10-28 à 15 12 02" src="https://user-images.githubusercontent.com/45975964/198601616-cfc7ca68-38c1-4036-9770-21683cdbe7d0.png">

**Comment by kevingilles** [comment]
Could this happen is I try to scan using altbeacon while one of my imported lib did the same?

**Comment by davidgyoung** [comment]
While it is possible that contributes to the cause, but I don't think that would fundamentally change anything.   Because the BeaconManger is a singleton it is shared by both your app and all libraries your app uses, so any problem like this might happen regardless of whether all calls are from your app or calls are also made from a dependent library.

From the stack traces I have seen so far, I don't understand why the threads are blocked.  There must be something wrong with my understanding about how `synchronized` is working in this case.
